Fix #101

dblsaiko opened the pull request:
This adds a canPlaceBlockOnSide check to block placement and a canPlaceBlockAt and canPlaceBlockOnSide check to part placement.
The part placement checks could theoretically be put into seperate methods (canPlacePartAt, canPlacePartOnSide?) in IMultipart, if you wish.

Commit:
Fix #101

amadornes commented:
Hmm... That'd be a good idea, yes. If you can do that, that'd be great :)

asiekierka commented:
I'll do that. Give me a bit.

Commit:
Added canPlacePartOnSide and canPlacePartAt to IMultipart

asiekierka commented on src/main/java/mcmultipart/api/multipart/IMultipart.java:
```diff
@@ -281,6 +281,14 @@ public default void onAdded(IPartInfo part) {
     }
 
+    public default boolean canPlacePartAt(IBlockAccess world, BlockPos pos) {
+        return true;
```
No, no, no! These should be proxying to the getBlock()'s original methods! But that's tricky, as the IPartInfo doesn't exist yet. I was writing a request to @amadornes to clarify what to do here, but you pushed this quicker... ah well.

dblsaiko replied:
hmm... I called IMultipart.getBlock() in the placement code, I imagine it's fine here too? But good point, they should query the Block methods. Whoops :P

asiekierka replied:
You *are* in the IMultipart, you can probably just call getBlock().

It may not be the finest way, but it's sufficient to proxy everything vanilla is asking of you.

dblsaiko replied:
Yeah, it seems to be working fine calling getBlock().canPlaceBlockAt(…)

asiekierka commented on src/main/java/mcmultipart/api/multipart/IMultipart.java:
```diff
@@ -281,6 +281,14 @@ public default void onAdded(IPartInfo part) {
+    }
+
+    public default boolean canPlacePartOnSide(IBlockAccess world, BlockPos pos, EnumFacing side) {
+        return canPlacePartAt(world, pos);
```
I wonder if this should even be proxied verbatim. I'd be in favor of passing both an EnumFacing and an IPartSlot here, not just an EnumFacing.

dblsaiko replied:
Yep, could be useful in some circumstances. I wonder, why not only have canPlacePartOnSide, since it's basically the same as canPlacePartAt, but with more information?

asiekierka replied:
I'd say keep both. You don't always have more information. That, or mark the added information as optional/not-necessary-to-provide somehow.

asiekierka commented:
Also, take a look at MultipartRegistry.registerStackWrapper, which calls canPlaceBlockAt but not canPlaceBlockOnSide.

Commit:
Made canPlacePartAt check Block's methods
Added canPlaceBlockOnSide check to stack wrapper
Pass IPartSlot to canPlacePartOnSide

asiekierka commented:
LGTM
